reactor-core/src/main/java/reactor/core/publisher/ContextPropagation.java

```diff
@@ -120,40 +121,34 @@ static Function<Context, Context> contextCapture(Predicate<Object> captureKeyPre
 		if (!isContextPropagationAvailable) {
 			return NO_OP;
 		}
-		return new ContextCaptureFunction(captureKeyPredicate, null);
+		return target -> ContextSnapshot.captureAllUsing(captureKeyPredicate, ContextRegistry.getInstance())
+			.updateContext(target);
 	}
 
-	static <T, R> BiConsumer<T, SynchronousSink<R>> contextRestoreForHandle(BiConsumer<T, SynchronousSink<R>> handler, CoreSubscriber<? super R> actual) {
-		if (!ContextPropagation.isContextPropagationAvailable() || actual.currentContext().isEmpty()) {
+	static <T, R> BiConsumer<T, SynchronousSink<R>> contextRestoreForHandle(BiConsumer<T, SynchronousSink<R>> handler, Supplier<Context> contextSupplier) {
```

Comment:
just a small comment that the supplier can be overhead for mono case since one extra object allocated ber subscriber

Reply:
I think this is fine since the `Supplier` in the hot path is `CoreSubscriber::currentContext` (ie. a method reference).